### ISSUE TITLE
Output is "delayed" by 2 cells

### ISSUE THREAD

**Issue by jmmcd** [brownan/sqlite3-kernel#1]
When I run some basic sqlite commands (e.g. `.tables`) and SQL commands (`SELECT`), the output appears to be correct, except that it's "delayed" by 2 cells -- see screenshot.

![sqlite_kernel_delay](https://cloud.githubusercontent.com/assets/167942/19642315/a845a356-99dc-11e6-979c-c2a5808ba5eb.png)

Here is version information from the Jupyter Notebook About menu: 

```
The version of the notebook server is 4.2.1 and is running on:
Python 3.5.1 |Anaconda 4.1.0 (x86_64)| (default, Jun 15 2016, 16:14:02) 
[GCC 4.2.1 Compatible Apple LLVM 4.2 (clang-425.0.28)]

Current Kernel Information:

3.13.0 2016-05-18 10:57:30 fc49f556e48970561d7ab6a2f24fdd7d9eb81ff2
```

**Comment by brownan** [comment]
Woah, that's weird. I wonder if there is some sort of local echo going on.

I don't know that much about how the jupyter kernel interface works—this project was a bit of an experiment—but I believe jupyter looks for the sqlite prompt for how to delimit the output from each command. It should also attempt to change the prompt so that it's unlikely to encounter the prompt in valid output.

Does your version of sqlite use a different prompt than the usual "sqlite> " prompt? If you're willing to do some debugging, I'd suggest putting some print or logging statements in the kernel code to see what output it's seeing and how it's delimiting, what prompt it set, etc.

Are you running this on a Mac? I wonder if there is some version incompatibility here.

For reference, my system works fine and is this version:

```
The version of the notebook server is 4.2.3-54ae42a and is running on:
Python 3.5.0+ (default, Oct 11 2015, 09:05:38) 
[GCC 5.2.1 20151010]

Current Kernel Information:

3.11.0 2016-02-15 17:29:24 3d862f207e3adc00f78066799ac5a8c282430a5f
```

(Ubuntu Linux 16.04)

**Comment by jmmcd** [comment]
Thanks for the reply. It is weird, isn't it!

My sqlite3 when started on the command line uses the typical `sqlite>` prompt.

Yes, I'm on Mac.

After looking at the kernel code, the first thing that comes to mind is that that stray semicolon in the output I showed looks like it's coming from a place where the kernel assumes it is seeing a "continuation".

I'll take another look later.

**Comment by brownan** [comment]
I'll see if I can get my hands on a mac and replicate this when I get some spare time. Let me know if you figure anything else out!

**Comment by jmmcd** [comment]
I investigated a bit more at a terminal and I see what is happening. As you speculated it's an extra echo for some reason.

```
In [13]: sql = Sqlite3REPL()
```

We see that after changing the prompt, our pattern is matching the `[PEXPECT_PROMPT>` in our prompt-change command:

```
In [14]: print(str(sql.child))
<pexpect.pty_spawn.spawn object at 0x106104128>
command: /Users/jmmcd/Documents/dev/anaconda/envs/py3k/bin/sqlite3
args: ['/Users/jmmcd/Documents/dev/anaconda/envs/py3k/bin/sqlite3']
searcher: None
buffer (last 100 chars): ' [PEXPECT_PROMPT+\r\n[PEXPECT_PROMPT>'
before (last 100 chars): '.prompt '
after: '[PEXPECT_PROMPT>'
match: '[PEXPECT_PROMPT>'
match_index: 0
[...]
```

We can confirm this -- the next command returns 1, i.e. matches the continuation prompt -- still in the buffer after our prompt-change command:

```
In [15]: sql._expect_prompt()
Out[15]: 1
```

If we `_expect_prompt()` again, we now finally match the _real_ prompt:

```
In [16]: sql._expect_prompt()
Out[16]: 0
```

And now things will work ok -- except that every command is doing this extra "echo":

```
In [17]: sql.run_command(".open student.db")
Out[17]: '.open student.db\r\n'
In [19]: sql.run_command("select * from student;")
Out[19]: 'select * from student;\r\nEnda Kenny|Education|1203\r\nMichael D. Higgins|Arts|1610\r\nMichael Noonan|Business|1705\r\nMichael Martin|Law|1750\r\n'
```

**Comment by HenryLeongStat** [comment]
I got the same problem. Did you find any solution?

**Comment by jnlaia** [comment]
Having the same problem in ubuntu 17.04, running jupyter in a docker container. jupyter version: 4.3.0

**Comment by chosak** [comment]
For those experiencing this issue, I encountered the same (on Mac) and found a workaround which I've opened as PR #3.

**Comment by chosak** [comment]
Can this issue be closed now that #3 has been merged in?

**Comment by brownan** [comment]
I assume so, since jmmcd already said the issue was fixed for them.
